File: _inc/laravel/app/Http/Controllers/Bills/LoanOptionController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    DatabaseConstants,
    MiddlewaresConstants,
    UsersConstants,
    ViewsConstants as VW
};
use App\Models\LoanOption;
use App\Traits\ChecksLogin;
use App\Traits\ChecksPermissions;
use Illuminate\Http\{Request, RedirectResponse, JsonResponse};
use Illuminate\Support\Facades\{Auth, DB, Log, Route, View as ViewFacade};
use Illuminate\View\View;

final class LoanOptionController extends Controller
{
    use ChecksLogin, ChecksPermissions;

    public function __construct()
    {
        $this->middleware(MiddlewaresConstants::AUTH);
    }

    public function index(Request $request): View|RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $action, $method, $class, $base) {
            Log::debug($class . '::' . $action . ' start', [UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
            if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
            if (($r = self::guard($request, 'manage loan option', VW::LN_OPT . '.index')) !== true) return $r;
            try {
                $qStart = microtime(true);
                $loanOptions = LoanOption::where(DatabaseConstants::TABLE_CREATOR, $request->user()?->creatorId() ?? null)->get();
                $this->logExecutionTime($qStart, $action, 'fetchLoanOptions');
                Log::info($class . '::' . $action . ' fetched', ['count' => $loanOptions->count(), UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
                $viewPath = VW::LN_OPT . '.' . $action;
                if (!ViewFacade::exists($viewPath)) {
                    Log::error($class . '::' . $action . ' missing view', ['view_path' => $viewPath]);
                    return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                }
                $renderStart = microtime(true);
                $resp = view($viewPath, compact('loanOptions'));
                $this->logExecutionTime($renderStart, $action, 'renderView');
                return $resp;
            } catch (\Throwable $e) {
                Log::error($class . '::' . $action . ' failed', ['error' => $e->getMessage(), UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function create(Request $request): View|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $action, $method, $class, $base) {
            Log::debug($class . '::' . $action . ' start', [UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
            if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
            if (($r = self::guard($request, 'create loan option', VW::LN_OPT . '.index')) !== true) return $r;
            $viewPath = VW::LN_OPT . '.' . $action;
            if (!ViewFacade::exists($viewPath)) {
                Log::error($class . '::' . $action . ' missing view', ['view_path' => $viewPath]);
                return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
            }
            $renderStart = microtime(true);
            $resp = view($viewPath);
            $this->logExecutionTime($renderStart, $action, 'renderView');
            return $resp;
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function store(Request $request): RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $action, $method, $class, $base) {
            Log::debug($class . '::' . $action . ' start', [UsersConstants::COL_USER_ID => $request->user()?->id ?? null, 'input' => $request->only('name')]);
            if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
            if (($r = self::guard($request, 'create loan option', VW::LN_OPT . '.index')) !== true) return $r;
            $valStart = microtime(true);
            $request->validate(['name' => 'required|string|max:20']);
            $this->logExecutionTime($valStart, $action, 'validateRequest');
            try {
                $txnStart = microtime(true);
                DB::transaction(function () use ($request, $class, $action) {
                    $crtStart = microtime(true);
                    $opt = LoanOption::create(['name' => $request->name, DatabaseConstants::TABLE_CREATOR => $request->user()?->creatorId() ?? null]);
                    $this->logExecutionTime($crtStart, $action, 'createLoanOption');
                    Log::info($class . '::' . $action . ' created', ['loan_option_id' => $opt->id ?? null, UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
                });
                $this->logExecutionTime($txnStart, $action, 'transaction');
                return redirect()->route(VW::LN_OPT . '.index')->with('success', __('Loan option successfully created.'));
            } catch (\Throwable $e) {
                Log::error($class . '::' . $action . ' failed', ['error' => $e->getMessage(), UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function edit(LoanOption $loanOption, Request $request): View|RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($loanOption, $request, $action, $method, $class, $base) {
            Log::debug($class . '::' . $action . ' start', [UsersConstants::COL_USER_ID => $request->user()?->id ?? null, 'loan_option_id' => $loanOption->id ?? null]);
            if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
            if (($r = self::guard($request, 'edit loan option', VW::LN_OPT . '.index')) !== true) return $r;
            if (!self::isOwner($loanOption)) {
                Log::warning($class . '::' . $action . ' forbidden', ['loan_option_id' => $loanOption->id ?? null, UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
                return defaultPermissionDenial($request, new \Exception('owner'), $class . '::' . $action);
            }
            $viewPath = VW::LN_OPT . '.' . $action;
            if (!ViewFacade::exists($viewPath)) {
                Log::error($class . '::' . $action . ' missing view', ['view_path' => $viewPath]);
                return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
            }
            $renderStart = microtime(true);
            $resp = view($viewPath, compact('loanOption'));
            $this->logExecutionTime($renderStart, $action, 'renderView');
            return $resp;
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'loan_option_id' => $loanOption->id ?? null]);
    }

    public function update(Request $request, LoanOption $loanOption): RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $loanOption, $action, $method, $class, $base) {
            Log::debug($class . '::' . $action . ' start', [UsersConstants::COL_USER_ID => $request->user()?->id ?? null, 'loan_option_id' => $loanOption->id ?? null, 'input' => $request->only('name')]);
            if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
            if (($r = self::guard($request, 'edit loan option', VW::LN_OPT . '.index')) !== true) return $r;
            if (!self::isOwner($loanOption)) {
                Log::warning($class . '::' . $action . ' forbidden', ['loan_option_id' => $loanOption->id ?? null, UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
                return defaultPermissionDenial($request, new \Exception('owner'), $class . '::' . $action);
            }
            $valStart = microtime(true);
            $request->validate(['name' => 'required|string|max:20']);
            $this->logExecutionTime($valStart, $action, 'validateRequest');
            try {
                $txnStart = microtime(true);
                DB::transaction(function () use ($request, $loanOption, $class, $action) {
                    $updStart = microtime(true);
                    $loanOption->update(['name' => $request->name]);
                    $this->logExecutionTime($updStart, $action, 'updateLoanOption');
                    Log::info($class . '::' . $action . ' updated', ['loan_option_id' => $loanOption->id ?? null, UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
                });
                $this->logExecutionTime($txnStart, $action, 'transaction');
                return redirect()->route(VW::LN_OPT . '.index')->with('success', __('Loan option successfully updated.'));
            } catch (\Throwable $e) {
                Log::error($class . '::' . $action . ' failed', ['error' => $e->getMessage(), 'loan_option_id' => $loanOption->id ?? null, UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'loan_option_id' => $loanOption->id ?? null]);
    }

    public function destroy(LoanOption $loanOption, Request $request): RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($loanOption, $request, $action, $method, $class, $base) {
            Log::debug($class . '::' . $action . ' start', [UsersConstants::COL_USER_ID => $request->user()?->id ?? null, 'loan_option_id' => $loanOption->id ?? null]);
            if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
            if (($r = self::guard($request, 'delete loan option', VW::LN_OPT . '.index')) !== true) return $r;
            if (!self::isOwner($loanOption)) {
                Log::warning($class . '::' . $action . ' forbidden', ['loan_option_id' => $loanOption->id ?? null, UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
                return defaultPermissionDenial($request, new \Exception('owner'), $class . '::' . $action);
            }
            try {
                $txnStart = microtime(true);
                DB::transaction(function () use ($loanOption, $request, $class, $action) {
                    $delStart = microtime(true);
                    $loanOption->delete();
                    $this->logExecutionTime($delStart, $action, 'deleteLoanOption');
                    Log::info($class . '::' . $action . ' deleted', ['loan_option_id' => $loanOption->id ?? null, UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
                });
                $this->logExecutionTime($txnStart, $action, 'transaction');
                return redirect()->route(VW::LN_OPT . '.index')->with('success', __('Loan option successfully deleted.'));
            } catch (\Throwable $e) {
                Log::error($class . '::' . $action . ' failed', ['error' => $e->getMessage(), 'loan_option_id' => $loanOption->id ?? null, UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'loan_option_id' => $loanOption->id ?? null]);
    }

    public function show(): RedirectResponse
    {
        return redirect()->route(VW::LN_OPT . '.index');
    }

    private static function deny(Request $request, string $permission): RedirectResponse|JsonResponse|null
    {
        Log::info(__CLASS__ . '::' . __FUNCTION__ . ' checking permission', [
            UsersConstants::COL_USER_ID    => $request->user()->id,
            'permission' => $permission,
        ]);
        if (!$request->user()->can($permission)) {
            Log::warning(__CLASS__ . '::' . __FUNCTION__ . ' denied', [
                UsersConstants::COL_USER_ID    => $request->user()->id,
                'permission' => $permission,
            ]);
            return defaultPermissionDenial(
                $request,
                new \Illuminate\Auth\Access\AuthorizationException($permission),
                __CLASS__ . '::' . __FUNCTION__
            );
        }
        Log::info(__CLASS__ . '::' . __FUNCTION__ . ' granted', [
            UsersConstants::COL_USER_ID    => $request->user()->id,
            'permission' => $permission,
        ]);
        return null;
    }

    private static function isOwner(LoanOption $opt): bool|RedirectResponse
    {
        if (
            ($userOrRedirect = self::_checkLogin())
            instanceof \Illuminate\Http\RedirectResponse
        )
            return $userOrRedirect;
        $user = $userOrRedirect;
        return $opt[DatabaseConstants::TABLE_CREATOR] === $user?->creatorId();
    }
}

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/bills/loan_option_controller.py
import logging
import inspect
from django.contrib import messages
from django.core.exceptions import PermissionDenied
from django.db import transaction
from django.http import HttpRequest, HttpResponse
from django.shortcuts import get_object_or_404, redirect, render
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from .._traits.controller import Controller
from ....Models.bills.loan_option import LoanOption

logger = logging.getLogger(__name__)

class LoanOptionController(Controller):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.middleware(['auth'])

    @classmethod
    def index(cls, request: HttpRequest) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('manage loan option'):
                raise PermissionDenied('User lacks permission: manage loan option')
            opts = LoanOption.objects.filter(created_by=request.user.creator_id())
            return render(request, 'loanoption/index.html', {'loanoptions': opts})
        except PermissionDenied as err:
            return default_permission_denial(request, err=err, ref=REF, logger=logger)
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @classmethod
    def create(cls, request: HttpRequest) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('create loan option'):
                raise PermissionDenied('User lacks permission: create loan option')
            return render(request, 'loanoption/create.html')
        except PermissionDenied as err:
            return default_permission_denial(request, err=err, ref=REF,
                                             logger=logger,
                                             json={'error': 'Permission denied.'})
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @classmethod
    @transaction.atomic
    def store(cls, request: HttpRequest) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('create loan option'):
                raise PermissionDenied('User lacks permission: create loan option')
            name = request.POST.get('name')
            if not name:
                messages.error(request, 'Name is required.')
                return redirect(get_redirect_url(request))
            lo = LoanOption(name=name, created_by=request.user.creator_id())
            lo.save()
            messages.success(request, 'LoanOption successfully created.')
            return redirect('loanoption_index')
        except PermissionDenied as err:
            return default_permission_denial(request, err=err, ref=REF, logger=logger)
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @classmethod
    def show(cls, request: HttpRequest, loanoption_id: int) -> HttpResponse:
        return redirect('loanoption_index')

    @classmethod
    def edit(cls, request: HttpRequest, loanoption_id: int) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('edit loan option'):
                raise PermissionDenied('User lacks permission: edit loan option')
            lo = get_object_or_404(LoanOption, pk=loanoption_id)
            if lo.created_by != request.user.creator_id():
                raise PermissionDenied('Permission denied.')
            return render(request, 'loanoption/edit.html', {'loanoption': lo})
        except PermissionDenied as err:
            return default_permission_denial(request, err=err, ref=REF,
                                             logger=logger,
                                             json={'error': 'Permission denied.'})
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @classmethod
    @transaction.atomic
    def update(cls, request: HttpRequest, loanoption_id: int) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('edit loan option'):
                raise PermissionDenied('User lacks permission: edit loan option')
            lo = get_object_or_404(LoanOption, pk=loanoption_id)
            if lo.created_by != request.user.creator_id():
                raise PermissionDenied('Permission denied.')
            name = request.POST.get('name')
            if not name:
                messages.error(request, 'Name is required.')
                return redirect(get_redirect_url(request))
            lo.name = name
            lo.save()
            messages.success(request, 'LoanOption successfully updated.')
            return redirect('loanoption_index')
        except PermissionDenied as err:
            return default_permission_denial(request, err=err, ref=REF, logger=logger)
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @classmethod
    def destroy(cls, request: HttpRequest, loanoption_id: int) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('delete loan option'):
                raise PermissionDenied('User lacks permission: delete loan option')
            lo = get_object_or_404(LoanOption, pk=loanoption_id)
            if lo.created_by != request.user.creator_id():
                raise PermissionDenied('Permission denied.')
            lo.delete()
            messages.success(request, 'LoanOption successfully deleted.')
            return redirect('loanoption_index')
        except PermissionDenied as err:
            return default_permission_denial(request, err=err, ref=REF, logger=logger)
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)
